wxAssertion error when closing braille or speech settings panels.

<!--
Please thoroughly read NVDA's wiki article on how to fill in this template, including how to provide the required files.
Issues may be closed if the required information is not present.
https://github.com/nvaccess/nvda/wiki/Github-issue-template-explanation-and-examples
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests. 
-->

### Steps to reproduce:
1. Open, for example the NVDA voice settings,
2. try to change something or nothing, it is based on your preferences and taste,
3. press enter to save, and/or escape to destroy your changes
### Actual behavior:
NVDA plays an error sound and produces the following exception, see log

```
ERROR - unhandled exception (09:33:35.263) - MainThread (7112):
wx._core.wxAssertionError: C++ assertion ""GetWindow() != 0"" failed at ..\..\src\common\wincmn.cpp(3919) in wxWindowAccessible::GetDescription(): 

The above exception was the direct cause of the following exception:
SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set
```

### Expected behavior:
NVDA should close this dialog without these types of exceptions and errors
### System configuration
#### NVDA installed/portable/running from source:
installed
#### NVDA version:
NVDA version alpha-22123,8665526b

#### Windows version:
Using Windows version 10.0.19042 workstation

#### Name and version of other software in use when reproducing the issue:
Not applicable
#### Other information about your system:

### Other questions
#### Does the issue still occur after restarting your computer?
yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Not applicable
#### If add-ons are disabled, is your problem still occurring?
Not applicable, NVDA core level
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Not applicable

This is reproducible with the error log

ERROR - unhandled exception (11:33:41.769) - MainThread (8416):

wx._core.wxAssertionError: C++ assertion ""GetWindow() != 0"" failed at ..\..\src\common\wincmn.cpp(3919) in wxWindowAccessible::GetDescription(): 

 

The above exception was the direct cause of the following exception:

 

SystemError: <class 'wx._core.WindowDestroyEvent'> returned a result with an error set

 

 

From: Łukasz Golonka ***@***.***> 
Sent: Friday, April 9, 2021 11:07 AM
To: nvaccess/nvda ***@***.***>
Cc: zstanecic ***@***.***>; Author ***@***.***>
Subject: Re: [nvaccess/nvda] When opening any NVDA dialog, then pressing enter or escape to dismiss it, NVDA crashes (#12220)

 

Is this still reproducible? For me it isn't but I was never able to reproduce this on my system.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/issues/12220#issuecomment-816539238> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACVCDE2RGWOP7JHLRGLIU2TTH27R3ANCNFSM4ZY4WN3A> .  <https://github.com/notifications/beacon/ACVCDE3OJNK45NUPXWGPCGTTH27R3A5CNFSM4ZY4WN3KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOGCVWMZQ.gif>

--- a/source/gui/settingsDialogs.py
+++ b/source/gui/settingsDialogs.py
@@ -237,8 +237,7 @@ class SettingsDialog(
 		Sub-classes may extend this method.
 		This base method should always be called to clean up the dialog.
 		"""
-		self.DestroyChildren()
-		self.Destroy()
+		self.DestroyLater()
 		self.SetReturnCode(wx.ID_OK)
 
 	def onCancel(self, evt):
@@ -246,8 +245,7 @@ class SettingsDialog(
 		Sub-classes may extend this method.
 		This base method should always be called to clean up the dialog.
 		"""
-		self.DestroyChildren()
-		self.Destroy()
+		self.DestroyLater()
 		self.SetReturnCode(wx.ID_CANCEL)
 
 	def onApply(self, evt):
@@ -404,7 +402,7 @@ class MultiCategorySettingsDialog(SettingsDialog):
 	"""
 
 	title=""
-	categoryClasses=[]
+	categoryClasses:typing.List[typing.Type[SettingsPanel]] = []
 
 	class CategoryUnavailableError(RuntimeError): pass
 
@@ -428,8 +426,9 @@ class MultiCategorySettingsDialog(SettingsDialog):
 		self.initialCategory = initialCategory
 		self.currentCategory = None
 		self.setPostInitFocus = None
-		# dictionary key is index of category in self.catList, value is the instance. Partially filled, check for KeyError
-		self.catIdToInstanceMap = {}
+		# dictionary key is index of category in self.catList, value is the instance.
+		# Partially filled, check for KeyError
+		self.catIdToInstanceMap:typing.Dict[int, SettingsPanel] = {}
 
 		super(MultiCategorySettingsDialog, self).__init__(
 			parent,
@@ -635,37 +634,42 @@ class MultiCategorySettingsDialog(SettingsDialog):
 		else:
 			evt.Skip()
 
-	def _doSave(self):
+	def _validateAllPanels(self):
+		"""Check if all panels are valid, and can be saved
+		@note: raises ValueError if a panel is not valid. See c{SettingsPanel.isValid}
+		"""
 		for panel in self.catIdToInstanceMap.values():
 			if panel.isValid() is False:
 				raise ValueError("Validation for %s blocked saving settings" % panel.__class__.__name__)
+
+	def _saveAllPanels(self):
 		for panel in self.catIdToInstanceMap.values():
 			panel.onSave()
+
+	def _notifyAllPanelsSaveOccurred(self):
 		for panel in self.catIdToInstanceMap.values():
 			panel.postSave()
 
-	def onOk(self,evt):
+	def _doSave(self):
 		try:
-			self._doSave()
+			self._validateAllPanels()
+			self._saveAllPanels()
+			self._notifyAllPanelsSaveOccurred()
 		except ValueError:
-			log.debugWarning("", exc_info=True)
+			log.debugWarning("Error while saving settings:", exc_info=True)
 			return
-		for panel in self.catIdToInstanceMap.values():
-			panel.Destroy()
+
+	def onOk(self,evt):
+		self._doSave()
 		super(MultiCategorySettingsDialog,self).onOk(evt)
 
 	def onCancel(self,evt):
 		for panel in self.catIdToInstanceMap.values():
 			panel.onDiscard()
-			panel.Destroy()
 		super(MultiCategorySettingsDialog,self).onCancel(evt)
 
 	def onApply(self,evt):
-		try:
-			self._doSave()
-		except ValueError:
-			log.debugWarning("", exc_info=True)
-			return
+		self._doSave()
 		super(MultiCategorySettingsDialog,self).onApply(evt)
 
 
@@ -1000,17 +1004,9 @@ class SpeechSettingsPanel(SettingsPanel):
 		super(SpeechSettingsPanel,self).onPanelDeactivated()
 
 	def onDiscard(self):
-		# Work around wxAssertion error #12220
-		# Manually destroying the ExpandoTextCtrl when the settings dialog is
-		# exited prevents the wxAssertion.
-		self.synthNameCtrl.Destroy()
 		self.voicePanel.onDiscard()
 
 	def onSave(self):
-		# Work around wxAssertion error #12220
-		# Manually destroying the ExpandoTextCtrl when the settings dialog is
-		# exited prevents the wxAssertion.
-		self.synthNameCtrl.Destroy()
 		self.voicePanel.onSave()
 
 class SynthesizerSelectionDialog(SettingsDialog):
@@ -3216,17 +3212,9 @@ class BrailleSettingsPanel(SettingsPanel):
 		super(BrailleSettingsPanel,self).onPanelDeactivated()
 
 	def onDiscard(self):
-		# Work around wxAssertion error #12220
-		# Manually destroying the ExpandoTextCtrl when the settings dialog is
-		# exited prevents the wxAssertion.
-		self.displayNameCtrl.Destroy()
 		self.brailleSubPanel.onDiscard()
 
 	def onSave(self):
-		# Work around wxAssertion error #12220
-		# Manually destroying the ExpandoTextCtrl when the settings dialog is
-		# exited prevents the wxAssertion.
-		self.displayNameCtrl.Destroy()
 		self.brailleSubPanel.onSave()
 
 

--- a/source/gui/settingsDialogs.py
+++ b/source/gui/settingsDialogs.py
@@ -402,7 +402,7 @@ class MultiCategorySettingsDialog(SettingsDialog):
 	"""
 
 	title=""
-	categoryClasses:typing.List[typing.Type[SettingsPanel]] = []
+	categoryClasses: typing.List[typing.Type[SettingsPanel]] = []
 
 	class CategoryUnavailableError(RuntimeError): pass
 
@@ -428,7 +428,7 @@ class MultiCategorySettingsDialog(SettingsDialog):
 		self.setPostInitFocus = None
 		# dictionary key is index of category in self.catList, value is the instance.
 		# Partially filled, check for KeyError
-		self.catIdToInstanceMap:typing.Dict[int, SettingsPanel] = {}
+		self.catIdToInstanceMap: typing.Dict[int, SettingsPanel] = {}
 
 		super(MultiCategorySettingsDialog, self).__init__(
 			parent,
@@ -659,7 +659,7 @@ class MultiCategorySettingsDialog(SettingsDialog):
 			log.debugWarning("Error while saving settings:", exc_info=True)
 			return
 
-	def onOk(self,evt):
+	def onOk(self, evt):
 		self._doSave()
 		super(MultiCategorySettingsDialog,self).onOk(evt)
 

